packages/builder/src/components/automation/SetupPanel/RowSelectorTypes.svelte

```diff
@@ -35,153 +36,197 @@
+  <TextArea
+    value={fieldData}
+    on:change={e =>
+      onChange({
+        row: {
+          [field]: e.detail,
+        },
+      })}
+  />
 {:else if schema.type === "json"}
   <span>
-    <Editor
-      editorHeight="150"
-      mode="json"
-      on:change={e => {
-        if (e.detail?.value !== value[field]) {
-          onChange(e, field, schema.type)
-        }
-      }}
-      value={value[field]}
-    />
+    <div class="field-wrap json-field">
+      <CodeEditor
+        value={fieldData}
+        on:change={e => {
+          onChange({
+            row: {
+              [field]: e.detail,
+            },
+          })
+        }}
+      />
+    </div>
   </span>
 {:else if schema.type === "link"}
   <LinkedRowSelector
-    linkedRows={value[field]}
+    linkedRows={fieldData}
     {schema}
-    on:change={e => onChange(e, field)}
+    on:change={e =>
+      onChange({
+        row: {
+          [field]: e.detail,
+        },
+      })}
     useLabel={false}
   />
 {:else if schema.type === "bb_reference" || schema.type === "bb_reference_single"}
   <LinkedRowSelector
-    linkedRows={value[field]}
+    linkedRows={fieldData}
     {schema}
     linkedTableId={"ta_users"}
-    on:change={e => onChange(e, field)}
+    on:change={e =>
+      onChange({
+        row: {
+          [field]: e.detail,
+        },
+      })}
     useLabel={false}
   />
 {:else if attachmentTypes.includes(schema.type)}
   <div class="attachment-field-container">
     <div class="toggle-container">
       <Toggle
-        value={useAttachmentBinding}
+        value={meta?.fields?.[field]?.useAttachmentBinding}
         text={"Use bindings"}
         size={"XS"}
-        on:change={e => handleToggleChange(field, e)}
+        on:change={e => {
+          onChange({
+            row: {
+              [field]: null,
+            },
+            meta: {
+              fields: {
+                [field]: {
+                  useAttachmentBinding: e.detail,
+                },
+              },
+            },
+          })
+        }}
       />
     </div>
-    {#if !useAttachmentBinding}
+
+    {#if !meta?.fields?.[field]?.useAttachmentBinding}
       <div class="attachment-field-spacing">
         <KeyValueBuilder
-          on:change={async e => {
-            onChange(
-              {
-                detail:
+          on:change={e => {
+            onChange({
+              row: {
+                [field]:
                   schema.type === FieldType.ATTACHMENT_SINGLE ||
                   schema.type === FieldType.SIGNATURE_SINGLE
                     ? e.detail.length > 0
                       ? {
                           url: e.detail[0].name,
                           filename: e.detail[0].value,
                         }
-                      : {}
+                      : null
                     : e.detail.map(({ name, value }) => ({
                         url: name,
                         filename: value,
                       })),
```

Comment:
NAB this `onChange` callback is a little bit gnarly with the multiple ternary statements etc... Would be a lot more readable to break it out into a function and not use all the conditionals but not a huge deal.